client/nginx.go

```diff
@@ -1199,149 +1252,283 @@ func determineStreamUpdates(updatedServers []StreamUpstreamServer, nginxServers
+		mu.Lock()
+		resolvers, err := client.GetResolversWithContext(initialCtx)
+		mu.Unlock()
+		if err != nil {
+			return fmt.Errorf("failed to get Resolvers: %w", err)
+		}
+		stats.Resolvers = *resolvers
+		return nil
+	})
+
+	initialGroup.Go(func() error {
+		mu.Lock()
+		httpLimitRequests, err := client.GetHTTPLimitReqsWithContext(initialCtx)
+		mu.Unlock()
+		if err != nil {
+			return fmt.Errorf("failed to get HTTPLimitRequests: %w", err)
+		}
+		stats.HTTPLimitRequests = *httpLimitRequests
+		return nil
+	})
+
+	initialGroup.Go(func() error {
+		mu.Lock()
+		httpLimitConnections, err := client.GetHTTPConnectionsLimitWithContext(initialCtx)
+		mu.Unlock()
+		if err != nil {
+			return fmt.Errorf("failed to get HTTPLimitConnections: %w", err)
+		}
+		stats.HTTPLimitConnections = *httpLimitConnections
+		return nil
+	})
+
+	initialGroup.Go(func() error {
+		mu.Lock()
+		workers, err := client.GetWorkersWithContext(initialCtx)
+		mu.Unlock()
+		if err != nil {
+			return fmt.Errorf("failed to get Workers: %w", err)
+		}
+		stats.Workers = workers
+		return nil
+	})
 
-	limitReqs, err := client.GetHTTPLimitReqs()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get stats: %w", err)
+	if err := initialGroup.Wait(); err != nil {
+		return nil, fmt.Errorf("error returned from contacting Plus API: %w", err)
 	}
 
-	limitConnsHTTP, err := client.GetHTTPConnectionsLimit()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get stats: %w", err)
-	}
+	// Process stream endpoints if they exist
+	if slices.Contains(stats.endpoints, "stream") {
+		availableStreamGroup, asgCtx := errgroup.WithContext(ctx)
 
-	workers, err := client.GetWorkers()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get stats: %w", err)
-	}
+		availableStreamGroup.Go(func() error {
+			mu.Lock()
+			streamEndpoints, err := client.GetAvailableStreamEndpointsWithContext(asgCtx)
+			mu.Unlock()
+			if err != nil {
+				return fmt.Errorf("failed to get available Stream Endpoints: %w", err)
+			}
+			stats.streamEndpoints = streamEndpoints
+			return nil
+		})
 
-	streamZones := &StreamServerZones{}
-	streamUpstreams := &StreamUpstreams{}
-	limitConnsStream := &StreamLimitConnections{}
-	var streamZoneSync *StreamZoneSync
+		if err := availableStreamGroup.Wait(); err != nil {
+			return nil, fmt.Errorf("no useful metrics found in stream stats: %w", err)
+		}
 
-	if slices.Contains(endpoints, "stream") {
-		streamEndpoints, err := client.GetAvailableStreamEndpoints()
-		if err != nil {
-			return nil, fmt.Errorf("failed to get stats: %w", err)
+		streamGroup, sgCtx := errgroup.WithContext(ctx)
+
+		if slices.Contains(stats.streamEndpoints, "server_zones") {
+			streamGroup.Go(func() error {
+				streamServerZones, err := client.GetStreamServerZonesWithContext(sgCtx)
+				if err != nil {
+					return fmt.Errorf("failed to get streamServerZones: %w", err)
+				}
+				mu.Lock()
+				stats.StreamServerZones = *streamServerZones
+				mu.Unlock()
```

Comment:
Should this lock be around the client call like the rest of the collections above? Or do all of these calls need to `defer mu.Unlock()` so the lock is in place for the entire goroutine?

Same comment for the rest of these.

Reply:
That's a good spot. Actually, the way it was written originally was slightly different. The locks are to prevent concurrent writes to the stats struct. I think parallel requests to the API should be ok. Generally it's probably safe enough as different fields under the struct are being populated by different calls. But the race detector detected concurrent access so I have updated the code to prevent concurrent writes to the stats structure and made the code consistent. Let me know if that explanation and the code updates satisfy your concerns?

Reply:
or maybe not based on the failures in the pipeline :)